### PR TITLE
Refactor GridTools::distributed_compute_point_locations()

### PR DESCRIPTION
**Pull request by peterrum**
Extracted from PR #11804. Only one code path is tested (`perform_handshake==false`); the other one will be tested by the follow-up PRs.

### CODE REVIEW

**Review by luca-heltai — APPROVED** [review]
This was a long due restructuring of the algorithm. Thanks for doing this!

File: include/deal.II/grid/grid_tools.h [hunk 1]
```diff
@@ -1001,7 +1009,98 @@ namespace GridTools
...
+    const std::vector<std::vector<BoundingBox<spacedim>>> &global_bboxes,
+    const double                                           tolerance = 1e-10);
+
+  namespace internal
+  {
+    /**
+     * Data structure returned by
+     * GridTools::internal::distributed_compute_point_locations(). It provides
+     * information to perform GridTools::distributed_compute_point_locations()
+     * and to setup the communication pattern within
```
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h:1021]:
```suggestion
     * and to set up the communication pattern within
```

File: include/deal.II/grid/grid_tools.h [hunk 2]
```diff
@@ -1001,7 +1009,98 @@ namespace GridTools
...
+     * Utilities::MPI::RemotePointEvaluation::reinit().
+     *
+     * @note The name of the fields are chosen with
+     *   Utilities::MPI::RemotePointEvaluation in mind. Here, quantities are
+     *   computed at specified arbitrary positioned (and even remote) points
```
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h:1026]:
```suggestion
     *   computed at specified arbitrary positioned points (and even on remote processes in the MPI universe)
```

File: include/deal.II/grid/grid_tools.h [hunk 3]
```diff
@@ -1001,7 +1009,98 @@ namespace GridTools
...
+        send_components;
+
+      /**
+       * Ranks to send to.
+       */
+      std::vector<unsigned int> send_ranks;
+
+      /**
+       * Pointers of ranges within a send buffer to be sent to the ranks
+       * specified by send_ranks. The size of the of the send buffer is given
```
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h:1054]:
```suggestion
       * specified by send_ranks. The size of the send buffer is given
```

File: include/deal.II/grid/grid_tools.h [hunk 4]
```diff
@@ -1001,7 +1009,98 @@ namespace GridTools
...
+      /**
+       * Pointers of ranges within a receive buffer that are filled by ranks
+       * specified by recv_ranks. The size of the receive buffer is given by
+       * recv_ptrs.back().
+       */
+      std::vector<unsigned int> recv_ptrs;
+    };
+
+    /**
+     * A function that filles DistributedComputePointLocationsInternal.
```
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h:1086]:
```suggestion
     * A function that fills DistributedComputePointLocationsInternal.
```

File: include/deal.II/grid/grid_tools.h [hunk 5]
```diff
@@ -1001,7 +1009,98 @@ namespace GridTools
...
+     * If the input argument @p perform_handshake is set to false only
+     * the fields are filled needed by
+     * GridTools::internal::distributed_compute_point_locations(). If the input
```
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h:1089]:
```suggestion
     * the fields needed by
     * GridTools::internal::distributed_compute_point_locations() are filled. If the input
```

**Review by kronbichler — APPROVED** [review]
Very nice, I only have some small comments on the comments.

**Comment by peterrum** [comment]
@kronbichler I have fixed the typos! Thanks!

**Review by kronbichler — APPROVED** [review]

**Pull request merged**